crates/accelerate/src/gate_direction.rs

```diff
@@ -0,0 +1,147 @@
+///     true iff all two-qubit gates comply with the target's coupling constraints
+#[pyfunction]
+#[pyo3(name = "check_gate_direction_target")]
+fn py_check_with_target(py: Python, dag: &DAGCircuit, target: &Target) -> PyResult<bool> {
+    let target_check = |inst: &PackedInstruction, op_args: &[Qubit]| -> bool {
+        let qargs = smallvec![
+            PhysicalQubit::new(op_args[0].0),
+            PhysicalQubit::new(op_args[1].0)
+        ];
+
+        target.instruction_supported(inst.op.name(), Some(&qargs))
+    };
+
+    check_gate_direction(py, dag, &target_check, None)
+}
+
+// The main routine for checking gate directionality.
+//
+// gate_complies: a function returning true iff the two-qubit gate direction complies with directionality constraints
+//
+// qubit_mapping: used for mapping the index of a given qubit within an instruction qargs vector to the corresponding qubit index of the
+//  original DAGCircuit the pass was called with. This mapping is required since control flow blocks are represented by nested DAGCircuit
+//  objects whose instruction qubit indices are relative to the parent DAGCircuit they reside in, thus when we recurse into nested DAGs, we need
+//  to carry the mapping context relative to the original DAG.
+//  When qubit_mapping is None, the identity mapping is assumed
+fn check_gate_direction<T>(
+    py: Python,
+    dag: &DAGCircuit,
+    gate_complies: &T,
+    qubit_mapping: Option<&[Qubit]>,
+) -> PyResult<bool>
+where
+    T: Fn(&PackedInstruction, &[Qubit]) -> bool,
+{
+    for node in dag.op_nodes(false) {
+        let NodeType::Operation(packed_inst) = &dag.dag[node] else {
+            panic!("PackedInstruction is expected");
+        };
+
+        let inst_qargs = dag.get_qargs(packed_inst.qubits);
+
+        if let OperationRef::Instruction(py_inst) = packed_inst.op.view() {
+            if py_inst.control_flow() {
+                let circuit_to_dag = imports::CIRCUIT_TO_DAG.get_bound(py); // TODO: Take out of the recursion
+                let py_inst = py_inst.instruction.bind(py);
+
+                for block in py_inst.getattr("blocks")?.iter()? {
+                    let inner_dag: DAGCircuit = circuit_to_dag.call1((block?,))?.extract()?;
+
+                    let block_ok = if let Some(mapping) = qubit_mapping {
+                        let mapping = inst_qargs // Create a temp mapping for the recursive call
+                            .iter()
+                            .map(|q| mapping[q.0 as usize])
+                            .collect::<Vec<Qubit>>();
+
+                        check_gate_direction(py, &inner_dag, gate_complies, Some(&mapping))?
+                    } else {
+                        check_gate_direction(py, &inner_dag, gate_complies, Some(inst_qargs))?
+                    };
+
+                    if !block_ok {
+                        return Ok(false);
+                    }
+                }
+            }
+        } else if inst_qargs.len() == 2
```

Comment:
This is technically a behavior change as you won't be checking 2 qubit non-control flow `Instruction` objects anymore and the previous python version did. I'm not sure if we should care or not because at least on real backends there aren't connectivity constraints on non-gates like this pass is trying to check, but I could conceive of a backend where it would matter. The simplest fix would be to just remove the `else` here and add a `continue` like this:

```suggestion
        if let OperationRef::Instruction(py_inst) = packed_inst.op.view() {
            if py_inst.control_flow() {
                let circuit_to_dag = imports::CIRCUIT_TO_DAG.get_bound(py); // TODO: Take out of the recursion
                let py_inst = py_inst.instruction.bind(py);

                for block in py_inst.getattr("blocks")?.iter()? {
                    let inner_dag: DAGCircuit = circuit_to_dag.call1((block?,))?.extract()?;

                    let block_ok = if let Some(mapping) = qubit_mapping {
                        let mapping = inst_qargs // Create a temp mapping for the recursive call
                            .iter()
                            .map(|q| mapping[q.0 as usize])
                            .collect::<Vec<Qubit>>();

                        check_gate_direction(py, &inner_dag, gate_complies, Some(&mapping))?
                    } else {
                        check_gate_direction(py, &inner_dag, gate_complies, Some(inst_qargs))?
                    };

                    if !block_ok {
                        return Ok(false);
                    }
                }
                continue;
            }
        } 
        if inst_qargs.len() == 2
```

Reply:
I was assuming we can skip it, but ok, that's a fair point, better keep the check more general